File: frontend/src/types/state.ts
/**
 * TypeScript mirrors of backend/app/schemas/state.py (docs/state-schema.md).
 * Keep in sync with the Pydantic models -- this is the single source of
 * truth for shapes returned by the API; components should import from here
 * rather than redeclaring inline types.
 */

export type ResearchMode = "auto" | "assisted" | "manual";

export type DecisionType =
  | "technology_selection"
  | "architecture"
  | "vendor"
  | "build_vs_buy"
  | "strategic"
  | "other";

export type Priority = "high" | "medium" | "low";

export interface ResearchRequest {
  id: string;
  question: string;
  constraints: string[];
  alternatives_hint: string[];
  criteria_hint: string[];
  mode: ResearchMode;
  requested_by: string;
  created_at: string;
}

export type ScoringMethod = "weighted_score" | "threshold" | "cost_benefit";
export type Direction = "maximize" | "minimize";

export interface DecisionCriteria {
  name: string;
  description: string;
  weight: number;
  direction: Direction;
  scoring_method: ScoringMethod;
}

export interface ResearchQuestion {
  id: string;
  text: string;
  dimension: string;
  priority: Priority;
}

export interface ResearchPlan {
  objective: string;
  decision_type: DecisionType;
  alternatives: string[];
  criteria: DecisionCriteria[];
  research_questions: ResearchQuestion[];
  required_evidence: string[];
  constraints: string[];
  assumptions: string[];
}

export type SourceType =
  | "official_docs"
  | "paper"
  | "government"
  | "standard"
  | "vendor_docs"
  | "blog"
  | "forum"
  | "news"
  | "internal_kb"
  | string;

export interface RetrievedDocument {
  source_id: string;
  title: string;
  url: string | null;
  source_type: SourceType;
  retrieved_at: string;
  relevant_passage: string;
  claims: string[];
  relevance_score: number;
}

export interface Source {
  id: string;
  title: string;
  url: string | null;
  publisher: string | null;
  author: string | null;
  published_at: string | null;
  retrieved_at: string;
  source_type: SourceType;
  credibility_score: number;
  content_hash: string;
}

export interface Claim {
  id: string;
  text: string;
  source_id: string;
  research_question_id: string;
}

export type ClaimStatus =
  | "VERIFIED"
  | "PARTIALLY_VERIFIED"
  | "CONTRADICTED"
  | "UNSUPPORTED"
  | "OUTDATED";

export interface ClaimVerification {
  claim_id: string;
  status: ClaimStatus;
  supporting_sources: string[];
  contradicting_sources: string[];
  confidence: number;
  explanation: string;
}

export type EvidenceType =
  | "quantitative"
  | "qualitative"
  | "benchmark"
  | "documentation"
  | "expert_analysis"
  | "empirical_observation"
  | "policy_regulatory"
  | "user_provided";

export type EvidenceStrength = "strong" | "moderate" | "weak";

export interface EvidenceItem {
  id: string;
  claim_id: string;
  evidence_text: string;
  source_id: string;
  evidence_type: EvidenceType;
  strength: EvidenceStrength;
  confidence: number;
  limitations: string | null;
}

export type ConflictType =
  | "genuine"
  | "context_mismatch"
  | "version_mismatch"
  | "temporal"
  | "workload_mismatch"
  | "environment_mismatch";

export type ResolutionStatus =
  | "unresolved"
  | "resolved_context"
  | "resolved_favor_a"
  | "resolved_favor_b"
  | "irreconcilable";

export interface Conflict {
  id: string;
  claim_a_id: string;
  claim_b_id: string;
  conflict_type: ConflictType;
  explanation: string;
  resolution_status: ResolutionStatus;
}

export type AssumptionOrigin =
  | "user_provided"
  | "evidence_backed"
  | "inferred"
  | "hypothetical";

export interface Assumption {
  id: string;
  text: string;
  origin: AssumptionOrigin;
  affects: string[];
}

export type RiskCategory =
  | "technical"
  | "financial"
  | "security"
  | "operational"
  | "regulatory"
  | "vendor"
  | "scalability"
  | "execution"
  | "unknown";

export type RiskLevel = "low" | "medium" | "high";

export interface Risk {
  id: string;
  category: RiskCategory;
  description: string;
  probability: RiskLevel;
  impact: RiskLevel;
  severity: number;
  mitigation: string;
  evidence_ids: string[];
}

export interface AlternativeScore {
  alternative: string;
  criterion: string;
  score: number;
  rationale: string;
  evidence_ids: string[];
}

export interface SensitivityResult {
  criterion: string;
  weight_delta: number;
  recommendation_changed: boolean;
  new_recommended: string | null;
}

export interface DecisionMatrix {
  criteria: DecisionCriteria[];
  scores: AlternativeScore[];
  weighted_totals: Record<string, number>;
  recommended: string;
  sensitivity: SensitivityResult[];
}

export type DebateRole = "advocate" | "critic";

export interface DebateNote {
  id: string;
  role: DebateRole;
  alternative: string | null;
  claim: string;
  rationale: string;
  evidence_ids: string[];
  target_note_id: string | null;
}

export interface VerificationResult {
  cycle: number;
  passed: boolean;
  failed_checks: string[];
  routed_to: string | null;
}

export interface AgentRunMeta {
  agent_name: string;
  trace_id: string;
  start_time: string;
  end_time: string;
  latency_ms: number;
  tokens: number;
  model: string;
  tool_calls: string[];
  errors: string[];
  confidence: number | null;
}

export type SessionStatus = "pending" | "running" | "completed" | "failed" | "cancelled";

export interface ExecutionMetadata {
  trace_id: string;
  session_id: string;
  started_at: string;
  completed_at: string | null;
  status: SessionStatus;
  agent_runs: AgentRunMeta[];
  total_tokens: number;
  total_cost_usd: number;
  verification_cycles_used: number;
}

export interface FinalReport {
  executive_summary: string;
  research_question: string;
  decision_context: string;
  alternatives: string[];
  criteria: DecisionCriteria[];
  key_findings: string[];
  evidence: EvidenceItem[];
  contradictions: Conflict[];
  comparative_analysis: DecisionMatrix;
  risk_analysis: Risk[];
  assumptions: Assumption[];
  decision_rationale: string;
  confidence: number;
  limitations: string[];
  sources: Source[];
  citations: Record<string, string>;
}

File: backend/app/schemas/state.py
"""Shared typed state, implementing docs/state-schema.md verbatim.

All inter-agent communication flows through `ResearchState`. No agent
receives or returns a bare dict. These are the Phase 0 schema contracts,
made concrete here so Phase 1's single-agent flow (and later phases'
LangGraph nodes) build against a stable, validated shape.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class ResearchRequest(BaseModel):
    id: UUID
    question: str
    constraints: list[str] = Field(default_factory=list)
    alternatives_hint: list[str] = Field(default_factory=list)
    criteria_hint: list[str] = Field(default_factory=list)
    mode: Literal["auto", "assisted", "manual"] = "auto"
    requested_by: str
    created_at: datetime


class DecisionCriteria(BaseModel):
    name: str
    description: str
    weight: float
    direction: Literal["maximize", "minimize"]
    scoring_method: Literal["weighted_score", "threshold", "cost_benefit"]


class ResearchQuestion(BaseModel):
    id: str
    text: str
    dimension: str
    priority: Literal["high", "medium", "low"]


class ResearchPlan(BaseModel):
    objective: str
    decision_type: Literal[
        "technology_selection", "architecture", "vendor", "build_vs_buy",
        "strategic", "other",
    ]
    alternatives: list[str]
    criteria: list[DecisionCriteria]
    research_questions: list[ResearchQuestion]
    required_evidence: list[str]
    constraints: list[str]
    assumptions: list[str]


class RetrievedDocument(BaseModel):
    source_id: str
    title: str
    url: str | None
    source_type: Literal[
        "official_docs", "paper", "government", "standard", "vendor_docs",
        "blog", "forum", "news", "internal_kb",
    ]
    retrieved_at: datetime
    relevant_passage: str
    claims: list[str]
    relevance_score: float


class Source(BaseModel):
    id: str
    title: str
    url: str | None
    publisher: str | None
    author: str | None
    published_at: datetime | None
    retrieved_at: datetime
    source_type: str
    credibility_score: float
    content_hash: str


class Claim(BaseModel):
    id: str
    text: str
    source_id: str
    research_question_id: str


class ClaimVerification(BaseModel):
    claim_id: str
    status: Literal[
        "VERIFIED", "PARTIALLY_VERIFIED", "CONTRADICTED", "UNSUPPORTED",
        "OUTDATED",
    ]
    supporting_sources: list[str]
    contradicting_sources: list[str]
    confidence: float
    explanation: str


class EvidenceItem(BaseModel):
    id: str
    claim_id: str
    evidence_text: str
    source_id: str
    evidence_type: Literal[
        "quantitative", "qualitative", "benchmark", "documentation",
        "expert_analysis", "empirical_observation", "policy_regulatory",
        "user_provided",
    ]
    strength: Literal["strong", "moderate", "weak"]
    confidence: float
    limitations: str | None = None


class Conflict(BaseModel):
    id: str
    claim_a_id: str
    claim_b_id: str
    conflict_type: Literal[
        "genuine", "context_mismatch", "version_mismatch", "temporal",
        "workload_mismatch", "environment_mismatch",
    ]
    explanation: str
    resolution_status: Literal[
        "unresolved", "resolved_context", "resolved_favor_a",
        "resolved_favor_b", "irreconcilable",
    ]


class Assumption(BaseModel):
    id: str
    text: str
    origin: Literal["user_provided", "evidence_backed", "inferred", "hypothetical"]
    affects: list[str] = Field(default_factory=list)


class Risk(BaseModel):
    id: str
    category: Literal[
        "technical", "financial", "security", "operational", "regulatory",
        "vendor", "scalability", "execution", "unknown",
    ]
    description: str
    probability: Literal["low", "medium", "high"]
    impact: Literal["low", "medium", "high"]
    severity: float
    mitigation: str
    evidence_ids: list[str] = Field(default_factory=list)


class AlternativeScore(BaseModel):
    alternative: str
    criterion: str
    score: float
    rationale: str
    evidence_ids: list[str] = Field(default_factory=list)


class SensitivityResult(BaseModel):
    criterion: str
    weight_delta: float
    recommendation_changed: bool
    new_recommended: str | None = None


class DecisionMatrix(BaseModel):
    criteria: list[DecisionCriteria]
    scores: list[AlternativeScore]
    weighted_totals: dict[str, float]
    recommended: str
    sensitivity: list[SensitivityResult] = Field(default_factory=list)


class DebateNote(BaseModel):
    """One entry from the Advocate/Critic pass (Phase 4, brief §5.6).

    `role` marks which pass produced it. A critic note attacks a specific
    advocate note via `target_note_id` where possible, and must always carry
    a non-empty `rationale` — the brief requires every criticism to be
    justified with evidence/reasoning, never a bare disagreement.
    """

    id: str
    role: Literal["advocate", "critic"]
    alternative: str | None = None  # None = applies across alternatives
    claim: str
    rationale: str
    evidence_ids: list[str] = Field(default_factory=list)
    target_note_id: str | None = None  # set on critic notes attacking an advocate note


class VerificationResult(BaseModel):
    cycle: int
    passed: bool
    failed_checks: list[str] = Field(default_factory=list)
    routed_to: str | None = None


class AgentRunMeta(BaseModel):
    agent_name: str
    trace_id: str
    start_time: datetime
    end_time: datetime
    latency_ms: int
    tokens: int
    model: str
    tool_calls: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    confidence: float | None = None


class ExecutionMetadata(BaseModel):
    trace_id: str
    session_id: UUID
    started_at: datetime
    completed_at: datetime | None = None
    status: Literal["pending", "running", "completed", "failed", "cancelled"]
    agent_runs: list[AgentRunMeta] = Field(default_factory=list)
    total_tokens: int = 0
    total_cost_usd: float = 0.0
    verification_cycles_used: int = 0


class FinalReport(BaseModel):
    executive_summary: str
    research_question: str
    decision_context: str
    alternatives: list[str]
    criteria: list[DecisionCriteria]
    key_findings: list[str]
    evidence: list[EvidenceItem]
    contradictions: list[Conflict]
    comparative_analysis: DecisionMatrix
    risk_analysis: list[Risk]
    assumptions: list[Assumption]
    decision_rationale: str
    confidence: float
    limitations: list[str]
    sources: list[Source]
    citations: dict[str, str] = Field(default_factory=dict)


class ResearchState(BaseModel):
    request: ResearchRequest
    plan: ResearchPlan | None = None
    retrieved_documents: list[RetrievedDocument] = Field(default_factory=list)
    sources: list[Source] = Field(default_factory=list)
    claims: list[Claim] = Field(default_factory=list)
    evidence: list[EvidenceItem] = Field(default_factory=list)
    verified_claims: list[ClaimVerification] = Field(default_factory=list)
    contradictions: list[Conflict] = Field(default_factory=list)
    debate_notes: list[DebateNote] = Field(default_factory=list)
    assumptions: list[Assumption] = Field(default_factory=list)
    risks: list[Risk] = Field(default_factory=list)
    decision_matrix: DecisionMatrix | None = None
    draft_report: FinalReport | None = None
    verification_results: list[VerificationResult] = Field(default_factory=list)
    final_report: FinalReport | None = None
    execution_metadata: ExecutionMetadata
